hill_climbing.cpp

```diff
@@ -0,0 +1,289 @@
+
+                tempname2 = savename + "_itimp_zones" + paddedRun + ".csv";
+                writename = fnames.outputdir + tempname2;
+                if ((Rfp = fopen(writename.c_str(), "w")) == NULL)
+                    displayErrorMessage("cannot create threshold trace file %s\n", writename.c_str());
+
+                fprintf(Rfp, "configuration");
+                for (int i = 0; i < puno; i++)
+                    fprintf(Rfp, ",%i", pu[i].id);
+                fprintf(Rfp, "\n0");
+
+                for (int i = 0; i < puno; i++)
+                    fprintf(Rfp, ",%i", R[i]);
+                fprintf(Rfp, "\n");
+
+                iRowCounter = 0;
+                if (fnames.itimptracerows == 0)
+                    iRowLimit = 0;
+                else
+                    iRowLimit = floor(puvalid / fnames.itimptracerows);
+
+            }
+
+            void append(int i, int puno, const scost& reserve, const scost& change, const vector<int>& R)
+            {
+                iRowCounter++;
+                if (iRowCounter > iRowLimit)
+                    iRowCounter = 1;
+
+                if (iRowCounter == 1)
+                {
+                    fprintf(Rfp, "%i", i);
+
+                    fprintf(ttfp, "%i,%f,%i,%f,%f,%f,%f\n"
+                                , i, reserve.total
+                                , reserve.pus, reserve.cost, reserve.connection, reserve.penalty,
+                                change.total); // i,costthresh,pus,cost,connection,penalty
+
+                    for (int j = 0; j < puno; j++)
+                        fprintf(Rfp, ",%i", R[j]);
+                }
+            }
+
+            ~ImportTraceSaver()
+            {
+                if(ttfp != nullptr)
+                    fclose(ttfp);
+                if(Rfp != nullptr)
+                    fclose(Rfp);
+            }
+        private:
+            FILE* ttfp, * Rfp;
+            string writename;
+            int iRowCounter, iRowLimit;
+    };
+
+    void initialiseHillClimbing(int puno, int spno, const vector<spustuff>& pu, const vector<sconnections>& connections, vector<sspecies>& spec,
+        const vector<spu>& SM, vector<spu_out>& SM_out, double cm,  int aggexist,
+        vector<int>& R, double prop, int clumptype, int irun, stringstream& logBuffer, rng_engine& rngEngine)
+    {
+        scost reserve = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+
+        initialiseReserve(prop, pu, R, rngEngine);
+
+        if (aggexist)
+            ClearClumps(spno, spec, pu, SM, SM_out);
+
+        computeReserveValue(puno, spno, R, pu, connections, SM, SM_out, cm, spec, aggexist, reserve, clumptype, logBuffer);
+
+    } 
+
+
+    // iteratively improves a planning unit solutions
+    // a descent algorithm un-reserves planning units that don't have a negative value when removed
+    void hill_climbing(int puno, int spno, const vector<spustuff>& pu, const vector<sconnections>& connections,
```

Comment:
could you explain how this is different from the existing iterative improvement method?

Reply:
1. "Iterative improvement" is generic term for any iteration algorithm.
2. Current iterative improvement is left "as is" to avoid introducing of unexpected differences.
3. Added automatic detection when it converges.
4. Hill climbing is equivalents of multiple iterations of   "Iterative improvement" (+ changes to logging).